src/natcap/invest/spec.py

```diff
@@ -0,0 +1,2042 @@
+    """
+    permissions: str = 'r'
+    type: typing.ClassVar[str] = 'file'
+
+    @timeout
+    def validate(self, filepath: str):
+        """Validate a file against the requirements for this input.
+
+        Args:
+            filepath (string): The filepath to validate.
+
+        Returns:
+            A string error message if an error was found.  ``None`` otherwise.
+        """
+        if not os.path.exists(filepath):
+            return get_message('FILE_NOT_FOUND')
+
+        for letter, mode, descriptor in (
+                ('r', os.R_OK, 'read'),
+                ('w', os.W_OK, 'write'),
+                ('x', os.X_OK, 'execute')):
+            if letter in self.permissions and not os.access(filepath, mode):
+                return get_message('NEED_PERMISSION_FILE').format(permission=descriptor)
+
+    @staticmethod
+    def format_column(col: pandas.Series, base_path: str) -> pandas.Series:
+        """Format a column of a pandas dataframe that contains FileInput values.
+
+        File path values are cast to `pandas.StringDtype`. Relative paths are
+        expanded to absolute paths relative to `base_path`. NA values remain NA.
+
+        Args:
+            col: Column of a pandas dataframe to format
+            base_path: Base path of the source CSV. Relative file path values
+                will be expanded relative to this base path.
+
+        Returns:
+            Transformed dataframe column
+        """
+        return col.apply(
+            lambda p: p if pandas.isna(p) else utils.expand_path(str(p).strip(), base_path)
+        ).astype(pandas.StringDtype())
+
+@dataclasses.dataclass
+class SingleBandRasterInput(FileInput):
+    """A single-band raster input, or parameter, of an invest model.
+
+    This represents a raster file input (all GDAL-supported raster file types
+    are allowed), where only the first band is needed.
+
+    Attributes:
+        band: An `Input` representing the type of data expected in the
+            raster's first and only band
+        projected: Defaults to None, indicating a projected (as opposed to
+            geographic) coordinate system is not required. Set to True if a
+            projected coordinate system is required.
+        projection_units: Defaults to None. If `projected` is `True`, and a
+            specific unit of projection (such as meters) is required, indicate
+            it here.
+    """
+    band: typing.Union[Input, None] = None
+    projected: typing.Union[bool, None] = None
+    projection_units: typing.Union[pint.Unit, None] = None
+    type: typing.ClassVar[str] = 'raster'
+
+    @timeout
+    def validate(self, filepath: str):
+        """Validate a raster file against the requirements for this input.
+
+        Args:
+            filepath (string): The filepath to validate.
+
+        Returns:
+            A string error message if an error was found.  ``None`` otherwise.
+        """
+        file_warning = FileInput.validate(self, filepath)
```

Comment:
In the `VectorInput` case you called `super().validate(filepath)` for this purpose. That seems like the more general solution, though they may be equivalent in all cases we have so far.

Reply:
You're right, this can be `super()`. I think this was left over from a multiple inheritance setup I tried.

Reply:
Hmm interesting. If it's working for `VectorInput.validate` to call `super().validate()` then shouldn't it be okay for `SingleBandRasterInput.validate` to do the same? I guess I see why `RasterOrVectorInput.validate` cannot call `super()`, but not sure I understand the others.

Reply:
I made `RasterOrVectorInput` a separate class that does not inherit from `SingleBandRasterInput` or `VectorInput` so that we don't have these issues.